linkerd/app/inbound/src/allow_ips.rs

```diff
@@ -0,0 +1,64 @@
+use linkerd_app_core::{
+    metrics::{error_metrics::LabelError, tcp_accept_errors},
+    svc,
+    transport::OrigDstAddr,
+    Error,
+};
+use std::{collections::HashSet, net::SocketAddr, sync::Arc};
+use thiserror::Error;
+
+#[derive(Clone, Debug)]
+pub(crate) struct AllowIps {
+    ips: Arc<HashSet<SocketAddr>>,
+}
+
+#[derive(Copy, Clone, Debug, Default)]
+pub(crate) struct LabelAcceptErrors(());
+
+#[derive(Clone, Debug, Error)]
+#[error("inbound connections are not allowed on this IP address ({ip})")]
+struct InvalidIp {
+    ip: SocketAddr,
+}
+
+impl<T> svc::stack::Predicate<T> for AllowIps
+where
+    T: svc::Param<OrigDstAddr>,
+{
+    type Request = T;
+
+    fn check(&mut self, target: T) -> Result<Self::Request, Error> {
+        // Allowlist not configured.
+        if self.ips.is_empty() {
+            return Ok(target);
+        }
+
+        let OrigDstAddr(ip) = target.param();
+        if self.ips.contains(&ip) {
+            return Ok(target);
+        }
+
+        tracing::warn!(%ip, allowed = ?self.ips, "IP address not in `LINKERD2_PROXY_INBOUND_IPS`");
+        Err(InvalidIp { ip }.into())
+    }
+}
+
+impl AllowIps {
+    pub(crate) fn new(ips: HashSet<SocketAddr>) -> Self {
+        if ips.is_empty() {
+            tracing::info!("`LINKERD_PROXY_INBOUND_IPS` allowlist not configured, allowing all target addresses");
+        } else {
+            tracing::info!(allowed = ?ips, "Only allowing connections targeting `LINKERD_PROXY_INBOUND_IPS`");
```

Comment:
```suggestion
            tracing::debug!(allowed = ?ips, "Only allowing connections targeting `LINKERD_PROXY_INBOUND_IPS`");
```

I don't think this is a user-facing concern worth logging on every startup.